File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        
        sort(nums.begin(), nums.end());

        for(int i = 0; i < nums.size(); ++i)
        {
            if(i > 0 && nums[i] == nums[i-1])
            {
                continue;
            }

            int left = i + 1;
            int right = nums.size()-1;

            while(left < right)
            {
                int sum = nums[left] + nums[right] + nums[i];
                if(sum > 0)
                {
                    --right;
                }
                else if(sum < 0)
                {
                    ++left;
                }
                else
                {
                    vector<int> make = {nums[left],nums[right],nums[i]};
                    res.push_back(make);
                    ++left;

                    while(nums[left] == nums[left - 1] && left < right)
                    {
                        ++left;
                    }
                }
            }
        }
        return res;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()

        res = []

        for i, val in enumerate(nums):

            if i > 0 and val == nums[i - 1]:
                continue

            left, right = i + 1, len(nums) - 1
            
            while right > left:
                threeSum = val + nums[left] + nums[right]
                if threeSum == 0:
                    res.append([val, nums[left], nums[right]])
                    left += 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
                elif threeSum > 0:
                    right -= 1
                else:
                    left += 1
        
        return res